### ISSUE TITLE
training resumption reverts to `bert-base-cased` rather than the provided checkpoint

### ISSUE THREAD

**Issue by JamesDeAntonis** [stanford-futuredata/ColBERT#43]
Hello again,

When I run training from a checkpoint (providing the `--resume` flag and a `--checkpoint` flag, I think the training (that occurs in the lines around [here](https://github.com/stanford-futuredata/ColBERT/blob/3d64bf1540dcc4f641c40a83ddc1615985480406/colbert/training/training.py#L96)) doesn't actually use the checkpoint, but instead starts from the un-finetuned `bert-base` that would be used if starting from scratch.

Is this intended?  What should I do if I would like to resume from the checkpoint?  Some evidence that I am right (I think I am, but I could be wrong) is provided below.  If I am right, I'd be happy to provide a pull request with a fix.  Thanks!

---

Code evidence: There are two models defined, `checkpoint` (the already partially finetuned model) and `colbert` (taken from a fresh `bert-base`).  It looks like the `checkpoint` is largely ignored while `colbert` is what gets all the training action.

Anecdotal evidence: I tried to resume training after 80000 batches of size 32 each, at which point my model was pretty far along in its training.  However, when I hit resume, I noticed that the loss rate started at a poor value and improved dramatically, as is typically the case at the fresh start of the finetuning process.

**Comment by okhat** [comment]
There are three relevant flags:

* --resume (from data point)
* --checkpoint (start from model parameters)
* --resume_optimizer (resume from optimizer parameters)

When you pass the first two, the model is in fact resumed but not the optimizer. This leads to seemingly high loss for a number of steps until the optimizer catches up.

**Comment by okhat** [comment]
The --resume_optimizer flag is NOT implemented in the public repository (here), but I have an internal implementation that I can merge if you need it. Let me know.

**Comment by JamesDeAntonis** [comment]
that would be great, thanks!

**Comment by okhat** [comment]
Great --- I think I can do this Friday or Saturday.

**Comment by wonderingalways** [comment]
Hey!
Is the above optimizer reload code merged? If so I am unable to see the changes?  If it wont be official merged can I get the internal implementation through mail?

**Comment by tdieu29** [comment]
Even after loading the optimizer state dict from my previous run, resuming training still did not give me the average loss that I ended up with from the previous run. @JamesDeAntonis If you resolved this problem, could you please share how you did it? Thanks a lot!

**Comment by tdieu29** [comment]
Found the solution! For anyone who might encounter this problem, what I did was saving train_loss (in addition to saving the model state dict, optimizer state dict and batch idx) when saving the checkpoints! You need the train_loss from the previous run when you resume training if you want the average loss from the current run to start where you left off.
